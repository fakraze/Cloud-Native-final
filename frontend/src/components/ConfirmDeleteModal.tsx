import React from 'react';
import { AlertTriangle, X } from 'lucide-react';

interface ConfirmDeleteModalProps {
  isOpen: boolean;
  onClose: () => void;
  onConfirm: () => void;
  title: string;
  message: string;
  itemName?: string;
  isLoading?: boolean;
}

const ConfirmDeleteModal: React.FC<ConfirmDeleteModalProps> = ({
  isOpen,
  onClose,
  onConfirm,
  title,
  message,
  itemName,
  isLoading = false
}) => {
  if (!isOpen) return null;

  return (
    <div className="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50 p-4">
      <div className="bg-white rounded-lg max-w-md w-full">
        {/* Header */}
        <div className="flex items-center justify-between p-6 border-b">
          <div className="flex items-center space-x-3">
            <div className="flex-shrink-0">
              <AlertTriangle className="h-6 w-6 text-red-600" />
            </div>
            <h2 className="text-lg font-semibold text-gray-900">{title}</h2>
          </div>
          <button
            onClick={onClose}
            className="text-gray-400 hover:text-gray-600"
            disabled={isLoading}
          >
            <X className="h-5 w-5" />
          </button>
        </div>

        {/* Content */}
        <div className="p-6">
          <p className="text-sm text-gray-600 mb-4">
            {message}
          </p>
          
          {itemName && (
            <div className="bg-gray-50 rounded-lg p-3 mb-4">
              <p className="text-sm font-medium text-gray-900">
                {itemName}
              </p>
            </div>
          )}

          <p className="text-sm text-red-600">
            This action cannot be undone.
          </p>
        </div>

        {/* Actions */}
        <div className="flex items-center justify-end space-x-3 p-6 border-t bg-gray-50 rounded-b-lg">
          <button
            type="button"
            onClick={onClose}
            className="btn-secondary"
            disabled={isLoading}
          >
            Cancel
          </button>
          <button
            type="button"
            onClick={onConfirm}
            className="bg-red-600 text-white px-4 py-2 rounded-md text-sm font-medium hover:bg-red-700 focus:outline-none focus:ring-2 focus:ring-red-500 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed"
            disabled={isLoading}
          >
            {isLoading ? (
              <div className="flex items-center space-x-2">
                <div className="animate-spin rounded-full h-4 w-4 border-b-2 border-white"></div>
                <span>Deleting...</span>
              </div>
            ) : (
              'Delete'
            )}
          </button>
        </div>
      </div>
    </div>
  );
};

export default ConfirmDeleteModal;
