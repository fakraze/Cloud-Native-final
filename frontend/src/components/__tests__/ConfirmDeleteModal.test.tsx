import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '../../test/test-utils';
import userEvent from '@testing-library/user-event';
import ConfirmDeleteModal from '../ConfirmDeleteModal';

describe('ConfirmDeleteModal', () => {
  const defaultProps = {
    isOpen: true,
    onClose: vi.fn(),
    onConfirm: vi.fn(),
    title: 'Delete Item',
    message: 'Are you sure you want to delete this item?',
    itemName: 'Test Item',
    isLoading: false,
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('Rendering', () => {
    it('should render modal when isOpen is true', () => {
      render(<ConfirmDeleteModal {...defaultProps} />);
      
      expect(screen.getByText('Delete Item')).toBeInTheDocument();
      expect(screen.getByText('Are you sure you want to delete this item?')).toBeInTheDocument();
      expect(screen.getByText('Test Item')).toBeInTheDocument();
    });

    it('should not render modal when isOpen is false', () => {
      render(<ConfirmDeleteModal {...defaultProps} isOpen={false} />);
      
      expect(screen.queryByText('Delete Item')).not.toBeInTheDocument();
    });

    it('should render without itemName when not provided', () => {
      const props = { ...defaultProps, itemName: undefined };
      render(<ConfirmDeleteModal {...props} />);
      
      expect(screen.getByText('Delete Item')).toBeInTheDocument();
      expect(screen.getByText('Are you sure you want to delete this item?')).toBeInTheDocument();
    });

    it('should render loading state correctly', () => {
      render(<ConfirmDeleteModal {...defaultProps} isLoading={true} />);
      
      const confirmButton = screen.getByText('Deleting...');
      const cancelButton = screen.getByText('Cancel');
      
      expect(confirmButton).toBeInTheDocument();
      expect(cancelButton).toBeDisabled();
    });
  });

  describe('User Interactions', () => {
    it('should call onClose when close button is clicked', async () => {
      const user = userEvent.setup();
      render(<ConfirmDeleteModal {...defaultProps} />);
      
      const buttons = screen.getAllByRole('button');
      const closeButton = buttons[0]; // The X button is the first button
      await user.click(closeButton);
      
      expect(defaultProps.onClose).toHaveBeenCalledTimes(1);
    });

    it('should call onClose when cancel button is clicked', async () => {
      const user = userEvent.setup();
      render(<ConfirmDeleteModal {...defaultProps} />);
      
      const cancelButton = screen.getByText('Cancel');
      await user.click(cancelButton);
      
      expect(defaultProps.onClose).toHaveBeenCalledTimes(1);
    });

    it('should call onConfirm when confirm button is clicked', async () => {
      const user = userEvent.setup();
      render(<ConfirmDeleteModal {...defaultProps} />);
      
      const confirmButton = screen.getByText('Delete');
      await user.click(confirmButton);
      
      expect(defaultProps.onConfirm).toHaveBeenCalledTimes(1);
    });

    it('should handle keyboard events correctly', async () => {
      const user = userEvent.setup();
      render(<ConfirmDeleteModal {...defaultProps} />);
      
      // Focus the modal first
      const modal = screen.getByText('Delete Item').closest('div');
      if (modal) {
        modal.focus();
      }
      
      // Test Escape key
      await user.keyboard('{Escape}');
      // Note: Escape key handling would need to be implemented in the component
    });

    it('should not allow interactions when loading', async () => {
      const user = userEvent.setup();
      render(<ConfirmDeleteModal {...defaultProps} isLoading={true} />);
      
      const cancelButton = screen.getByText('Cancel');
      
      expect(cancelButton).toBeDisabled();
    });
  });

  describe('Accessibility', () => {
    it('should have proper aria labels and structure', () => {
      render(<ConfirmDeleteModal {...defaultProps} />);
      
      expect(screen.getByText('Delete Item')).toBeInTheDocument();
      expect(screen.getByText('Are you sure you want to delete this item?')).toBeInTheDocument();
      expect(screen.getByText('This action cannot be undone.')).toBeInTheDocument();
    });
  });

  describe('Snapshots', () => {
    it('should match snapshot in default state', () => {
      const { container } = render(<ConfirmDeleteModal {...defaultProps} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot in loading state', () => {
      const { container } = render(<ConfirmDeleteModal {...defaultProps} isLoading={true} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot without itemName', () => {
      const props = { ...defaultProps, itemName: undefined };
      const { container } = render(<ConfirmDeleteModal {...props} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
