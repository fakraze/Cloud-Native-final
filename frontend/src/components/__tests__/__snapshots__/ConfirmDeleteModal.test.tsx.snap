// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConfirmDeleteModal > Snapshots > should match snapshot in default state 1`] = `
<div
  class="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50 p-4"
>
  <div
    class="bg-white rounded-lg max-w-md w-full"
  >
    <div
      class="flex items-center justify-between p-6 border-b"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="flex-shrink-0"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-triangle-alert h-6 w-6 text-red-600"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m21.73 18-8-14a2 2 0 0 0-3.48 0l-8 14A2 2 0 0 0 4 21h16a2 2 0 0 0 1.73-3"
            />
            <path
              d="M12 9v4"
            />
            <path
              d="M12 17h.01"
            />
          </svg>
        </div>
        <h2
          class="text-lg font-semibold text-gray-900"
        >
          Delete Item
        </h2>
      </div>
      <button
        class="text-gray-400 hover:text-gray-600"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-x h-5 w-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
      </button>
    </div>
    <div
      class="p-6"
    >
      <p
        class="text-sm text-gray-600 mb-4"
      >
        Are you sure you want to delete this item?
      </p>
      <div
        class="bg-gray-50 rounded-lg p-3 mb-4"
      >
        <p
          class="text-sm font-medium text-gray-900"
        >
          Test Item
        </p>
      </div>
      <p
        class="text-sm text-red-600"
      >
        This action cannot be undone.
      </p>
    </div>
    <div
      class="flex items-center justify-end space-x-3 p-6 border-t bg-gray-50 rounded-b-lg"
    >
      <button
        class="btn-secondary"
        type="button"
      >
        Cancel
      </button>
      <button
        class="bg-red-600 text-white px-4 py-2 rounded-md text-sm font-medium hover:bg-red-700 focus:outline-none focus:ring-2 focus:ring-red-500 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed"
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`ConfirmDeleteModal > Snapshots > should match snapshot in loading state 1`] = `
<div
  class="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50 p-4"
>
  <div
    class="bg-white rounded-lg max-w-md w-full"
  >
    <div
      class="flex items-center justify-between p-6 border-b"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="flex-shrink-0"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-triangle-alert h-6 w-6 text-red-600"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m21.73 18-8-14a2 2 0 0 0-3.48 0l-8 14A2 2 0 0 0 4 21h16a2 2 0 0 0 1.73-3"
            />
            <path
              d="M12 9v4"
            />
            <path
              d="M12 17h.01"
            />
          </svg>
        </div>
        <h2
          class="text-lg font-semibold text-gray-900"
        >
          Delete Item
        </h2>
      </div>
      <button
        class="text-gray-400 hover:text-gray-600"
        disabled=""
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-x h-5 w-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
      </button>
    </div>
    <div
      class="p-6"
    >
      <p
        class="text-sm text-gray-600 mb-4"
      >
        Are you sure you want to delete this item?
      </p>
      <div
        class="bg-gray-50 rounded-lg p-3 mb-4"
      >
        <p
          class="text-sm font-medium text-gray-900"
        >
          Test Item
        </p>
      </div>
      <p
        class="text-sm text-red-600"
      >
        This action cannot be undone.
      </p>
    </div>
    <div
      class="flex items-center justify-end space-x-3 p-6 border-t bg-gray-50 rounded-b-lg"
    >
      <button
        class="btn-secondary"
        disabled=""
        type="button"
      >
        Cancel
      </button>
      <button
        class="bg-red-600 text-white px-4 py-2 rounded-md text-sm font-medium hover:bg-red-700 focus:outline-none focus:ring-2 focus:ring-red-500 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed"
        disabled=""
        type="button"
      >
        <div
          class="flex items-center space-x-2"
        >
          <div
            class="animate-spin rounded-full h-4 w-4 border-b-2 border-white"
          />
          <span>
            Deleting...
          </span>
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`ConfirmDeleteModal > Snapshots > should match snapshot without itemName 1`] = `
<div
  class="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50 p-4"
>
  <div
    class="bg-white rounded-lg max-w-md w-full"
  >
    <div
      class="flex items-center justify-between p-6 border-b"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="flex-shrink-0"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-triangle-alert h-6 w-6 text-red-600"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m21.73 18-8-14a2 2 0 0 0-3.48 0l-8 14A2 2 0 0 0 4 21h16a2 2 0 0 0 1.73-3"
            />
            <path
              d="M12 9v4"
            />
            <path
              d="M12 17h.01"
            />
          </svg>
        </div>
        <h2
          class="text-lg font-semibold text-gray-900"
        >
          Delete Item
        </h2>
      </div>
      <button
        class="text-gray-400 hover:text-gray-600"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-x h-5 w-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
      </button>
    </div>
    <div
      class="p-6"
    >
      <p
        class="text-sm text-gray-600 mb-4"
      >
        Are you sure you want to delete this item?
      </p>
      <p
        class="text-sm text-red-600"
      >
        This action cannot be undone.
      </p>
    </div>
    <div
      class="flex items-center justify-end space-x-3 p-6 border-t bg-gray-50 rounded-b-lg"
    >
      <button
        class="btn-secondary"
        type="button"
      >
        Cancel
      </button>
      <button
        class="bg-red-600 text-white px-4 py-2 rounded-md text-sm font-medium hover:bg-red-700 focus:outline-none focus:ring-2 focus:ring-red-500 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed"
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
